### PR TITLE
improvement: Force closing thread if file watching cancel hanged

### PR DESCRIPTION
**Pull request by tgodzik**
I got a frozen Metals instance and it seems to have been stuck on cancel:

```
java.lang.Thread.State: WAITING (on object monitor)
at java.lang.Object.wait(java.base@17.0.2/Native Method)
- waiting on
at java.lang.Object.wait(java.base@17.0.2/Object.java:338)
at sun.nio.fs.AbstractPoller$Request.awaitResult(java.base@17.0.2/AbstractPoller.java:201)
- locked <0x00000000d7e62158> (a sun.nio.fs.AbstractPoller$Request)
at sun.nio.fs.AbstractPoller.invoke(java.base@17.0.2/AbstractPoller.java:230)
at sun.nio.fs.AbstractPoller.cancel(java.base@17.0.2/AbstractPoller.java:138)
at sun.nio.fs.LinuxWatchService$LinuxWatchKey.cancel(java.base@17.0.2/LinuxWatchService.java:136)
at com.swoval.files.NioPathWatcherService$CachedWatchDirectory.close(NioPathWatcherService.java:126)
at com.swoval.files.NioPathWatcherService.close(NioPathWatcherService.java:182)
at com.swoval.files.NioPathWatcher.close(NioPathWatcher.java:366)
at com.swoval.files.SymlinkFollowingPathWatcher.close(SymlinkFollowingPathWatcher.java:133)
at scala.meta.internal.metals.watcher.FileWatcher$.$anonfun$startWatch$7(FileWatcher.scala:199)
at scala.meta.internal.metals.watcher.FileWatcher$$$Lambda$4991/0x0000000801a10000.apply$mcV$sp(Unknown Source)
at scala.meta.internal.metals.watcher.FileWatcher.cancel(FileWatcher.scala:50)
at scala.meta.internal.metals.Cancelable$.$anonfun$cancelAll$1(Cancelable.scala:29)
at scala.meta.internal.metals.Cancelable$$$Lambda$3649/0x00000008016433f8.apply(Unknown Source)
at scala.collection.immutable.List.foreach(List.scala:333)
at scala.meta.internal.metals.Cancelable$.cancelAll(Cancelable.scala:28)
at scala.meta.internal.metals.MutableCancelable.cancel(MutableCancelable.scala:25)
at scala.meta.internal.metals.MetalsLspService.cancel(MetalsLspService.scala:148)
at scala.meta.internal.metals.MetalsLspService.shutdown(MetalsLspService.scala:1051)
at scala.meta.metals.MetalsLanguageServer.shutdown(MetalsLanguageServer.scala:221)
at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.2/Native Method)
at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.2/NativeMethodAccessorImpl.java:77)
at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.2/DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(java.base@17.0.2/Method.java:568)
at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint$$Lambda$120/0x0000000800d93d78.apply(Unknown Source)
at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.jav
```

Let's try and interrupt the thread if it's taking too long to close. AbstractPoller checks for interrupted.

### CODE REVIEW

**Comment by tgodzik** [comment]
Opinions? Not sure if this is the best way to handle it.

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala [hunk 1]
```diff
@@ -196,9 +205,20 @@ object FileWatcher {
...
+        hanged.set(Thread.currentThread())
+        watcher.close()
+        stopWatchingSignal.set(true)
+        thread.interrupt()
+        hanged.set(null)
+      }
+      // watcher.close() might hang so we don't want to get stuck here
+      Try(Await.ready(closing, 3.seconds)) match {
+        case Failure(_: TimeoutException) =>
+          Option(hanged.get()).foreach(_.interrupt())
```
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:219]: Probably it's also needed to call `stopWatchingSignal.set(true)`  to stop `metals-watch-callback-thread`

**Review by dos65 — APPROVED** [review]
I never experienced such issue but I'm fine with this approach if it works

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala (hunk 1 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:219]: I moved `stopWatchingSignal.set(true)` before close which I found to hang

**Comment by tgodzik** [comment]
Anyone against merging this?

**Review by tanishiking — APPROVED** [review]
LGTM, from my side, let's see how it goes :+1:

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala (hunk 1 above)
Comment (dos65) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:219]: oh, now i got it

**Pull request merged**